Add static code analysis with PHPStan level 0

Art4 opened the pull request:
This PR adds [PHPStan](https://phpstan.org/) as static code analysis tool and adds a new composer script `phpstan`. Using this script PHPStan on [level 0](https://phpstan.org/user-guide/rule-levels) founds 67 errors, that will be fixed in this PR.

<details><summary>Details</summary>

```bash

 $ bin/composer.phar run phpstan
Xdebug: [Step Debug] Could not connect to debugging client. Tried: localhost:9000 (through xdebug.client_host/xdebug.client_port).
> phpstan analyze --memory-limit 1024M --configuration .phpstan.neon
Xdebug: [Step Debug] Could not connect to debugging client. Tried: localhost:9000 (through xdebug.client_host/xdebug.client_port).
 1014/1014 [▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓] 100%

 ------ ------------------------------------------------------------------------------ 
  Line   App/Mode.php                                                                  
 ------ ------------------------------------------------------------------------------ 
  :213   Class Friendica\App\Mode referenced with incorrect case: Friendica\App\MODE.  
 ------ ------------------------------------------------------------------------------ 

 ------ ----------------------------------------------------------------------------------- 
  Line   BaseCollection.php                                                                 
 ------ ----------------------------------------------------------------------------------- 
  :92    Unsafe usage of new static().                                                      
         💡 See: https://phpstan.org/blog/solving-phpstan-error-unsafe-usage-of-new-static  
  :105   Unsafe usage of new static().                                                      
         💡 See: https://phpstan.org/blog/solving-phpstan-error-unsafe-usage-of-new-static  
  :115   Unsafe usage of new static().                                                      
         💡 See: https://phpstan.org/blog/solving-phpstan-error-unsafe-usage-of-new-static  
  :131   Unsafe usage of new static().                                                      
         💡 See: https://phpstan.org/blog/solving-phpstan-error-unsafe-usage-of-new-static  
 ------ ----------------------------------------------------------------------------------- 

 ------ ------------------------------------------------------------------------------------------------ 
  Line   Console/Addon.php                                                                               
 ------ ------------------------------------------------------------------------------------------------ 
  :146   Method Friendica\Console\Addon::list() should return bool|int but return statement is missing.  
 ------ ------------------------------------------------------------------------------------------------ 

 ------ --------------------------------------------------------------------- 
  Line   Console/AutomaticInstallation.php                                    
 ------ --------------------------------------------------------------------- 
  :137   Instantiated class Friendica\Console\Exception not found.            
         💡 Learn more at https://phpstan.org/user-guide/discovering-symbols  
 ------ --------------------------------------------------------------------- 

 ------ ---------------------------------------------------------------------------------------------------- 
  Line   Console/Contact.php                                                                                 
 ------ ---------------------------------------------------------------------------------------------------- 
  :159   Method Friendica\Console\Contact::addContact() should return bool but return statement is missing.  
 ------ ---------------------------------------------------------------------------------------------------- 

 ------ --------------------------------------------------------------------------------------------- 
  Line   Console/FixAPDeliveryWorkerTaskParameters.php                                                
 ------ --------------------------------------------------------------------------------------------- 
  :71    Access to an undefined property Friendica\Console\FixAPDeliveryWorkerTaskParameters::$l10n.  
         💡 Learn more: https://phpstan.org/blog/solving-phpstan-access-to-undefined-property         
 ------ --------------------------------------------------------------------------------------------- 

 ------ ------------------------------------------------------------------------------------------------- 
  Line   Console/User.php                                                                                 
 ------ ------------------------------------------------------------------------------------------------- 
  :461   Method Friendica\Console\User::configUser() should return bool but return statement is missing.  
 ------ ------------------------------------------------------------------------------------------------- 

 ------ ----------------------------------------------------------------------------------- 
  Line   Content/Post/Entity/PostMedia.php                                                  
 ------ ----------------------------------------------------------------------------------- 
  :231   Unsafe usage of new static().                                                      
         💡 See: https://phpstan.org/blog/solving-phpstan-error-unsafe-usage-of-new-static  
  :258   Unsafe usage of new static().                                                      
         💡 See: https://phpstan.org/blog/solving-phpstan-error-unsafe-usage-of-new-static  
 ------ ----------------------------------------------------------------------------------- 

 ------ ------------------------------------------------------------------------------------------------------------------- 
  Line   Content/Widget/ContactBlock.php                                                                                    
 ------ ------------------------------------------------------------------------------------------------------------------- 
  :33    Template type widget of method Friendica\Content\Widget\ContactBlock::getHTML() is not referenced in a parameter.  
 ------ ------------------------------------------------------------------------------------------------------------------- 

 ------ ------------------------------------------------------------------------------------------------------------ 
  Line   Content/Widget/VCard.php                                                                                    
 ------ ------------------------------------------------------------------------------------------------------------ 
  :35    Template type widget of method Friendica\Content\Widget\VCard::getHTML() is not referenced in a parameter.  
 ------ ------------------------------------------------------------------------------------------------------------ 

 ------ ------------------------------------------------------------------ 
  Line   Core/Cache/Type/APCuCache.php                                     
 ------ ------------------------------------------------------------------ 
  :47    Class APCuIterator referenced with incorrect case: APCUIterator.  
  :128   Class APCuIterator referenced with incorrect case: APCUIterator.  
 ------ ------------------------------------------------------------------ 

 ------- ------------------------------------------------------ 
  Line    Model/Contact.php                                     
 ------- ------------------------------------------------------ 
  :2811   Cannot unset offset 'avatar' on array{photo: mixed}.  
 ------- ------------------------------------------------------ 

 ------- -------------------------- 
  Line    Model/Item.php            
 ------- -------------------------- 
  :3204   Undefined variable: $set  
 ------- -------------------------- 

 ------ -------------------------------- 
  Line   Module/Api/Mastodon/Search.php  
 ------ -------------------------------- 
  :161   Undefined variable: $since_id   
 ------ -------------------------------- 

 ------ ------------------------------------------------------------------------------------------------------------------------------------------------------------------ 
  Line   Module/Debug/ActivityPubConversion.php                                                                                                                            
 ------ ------------------------------------------------------------------------------------------------------------------------------------------------------------------ 
  :26    Inner named functions are not supported by PHPStan. Consider refactoring to an anonymous function, class method, or a top-level-defined function. See issue #165  
         (https://github.com/phpstan/phpstan/issues/165) for more details.                                                                                                 
  :46    Function visible_whitespace not found.                                                                                                                            
         💡 Learn more at https://phpstan.org/user-guide/discovering-symbols                                                                                               
  :50    Function visible_whitespace not found.                                                                                                                            
         💡 Learn more at https://phpstan.org/user-guide/discovering-symbols                                                                                               
  :58    Function visible_whitespace not found.                                                                                                                            
         💡 Learn more at https://phpstan.org/user-guide/discovering-symbols                                                                                               
  :96    Undefined variable: $body                                                                                                                                         
  :110   Function visible_whitespace not found.                                                                                                                            
         💡 Learn more at https://phpstan.org/user-guide/discovering-symbols                                                                                               
  :117   Function visible_whitespace not found.                                                                                                                            
         💡 Learn more at https://phpstan.org/user-guide/discovering-symbols                                                                                               
 ------ ------------------------------------------------------------------------------------------------------------------------------------------------------------------ 

 ------ ------------------------------------------------------------------------------------------------------------------------------------------------------------------ 
  Line   Module/Debug/Babel.php                                                                                                                                            
 ------ ------------------------------------------------------------------------------------------------------------------------------------------------------------------ 
  :32    Inner named functions are not supported by PHPStan. Consider refactoring to an anonymous function, class method, or a top-level-defined function. See issue #165  
         (https://github.com/phpstan/phpstan/issues/165) for more details.                                                                                                 
  :45    Function visible_whitespace not found.                                                                                                                            
         💡 Learn more at https://phpstan.org/user-guide/discovering-symbols                                                                                               
  :51    Function visible_whitespace not found.                                                                                                                            
         💡 Learn more at https://phpstan.org/user-guide/discovering-symbols                                                                                               
  :57    Function visible_whitespace not found.                                                                                                                            
         💡 Learn more at https://phpstan.org/user-guide/discovering-symbols                                                                                               
  :62    Function visible_whitespace not found.                                                                                                                            
         💡 Learn more at https://phpstan.org/user-guide/discovering-symbols                                                                                               
  :73    Function visible_whitespace not found.                                                                                                                            
         💡 Learn more at https://phpstan.org/user-guide/discovering-symbols                                                                                               
  :79    Function visible_whitespace not found.                                                                                                                            
         💡 Learn more at https://phpstan.org/user-guide/discovering-symbols                                                                                               
  :85    Function visible_whitespace not found.                                                                                                                            
         💡 Learn more at https://phpstan.org/user-guide/discovering-symbols                                                                                               
  :95    Function visible_whitespace not found.                                                                                                                            
         💡 Learn more at https://phpstan.org/user-guide/discovering-symbols                                                                                               
  :101   Function visible_whitespace not found.                                                                                                                            
         💡 Learn more at https://phpstan.org/user-guide/discovering-symbols                                                                                               
  :109   Function visible_whitespace not found.                                                                                                                            
         💡 Learn more at https://phpstan.org/user-guide/discovering-symbols                                                                                               
  :113   Function visible_whitespace not found.                                                                                                                            
         💡 Learn more at https://phpstan.org/user-guide/discovering-symbols                                                                                               
  :119   Function visible_whitespace not found.                                                                                                                            
         💡 Learn more at https://phpstan.org/user-guide/discovering-symbols                                                                                               
  :124   Function visible_whitespace not found.                                                                                                                            
         💡 Learn more at https://phpstan.org/user-guide/discovering-symbols                                                                                               
  :135   Function visible_whitespace not found.                                                                                                                            
         💡 Learn more at https://phpstan.org/user-guide/discovering-symbols                                                                                               
  :144   Function visible_whitespace not found.                                                                                                                            
         💡 Learn more at https://phpstan.org/user-guide/discovering-symbols                                                                                               
  :150   Function visible_whitespace not found.                                                                                                                            
         💡 Learn more at https://phpstan.org/user-guide/discovering-symbols                                                                                               
  :161   Function visible_whitespace not found.                                                                                                                            
         💡 Learn more at https://phpstan.org/user-guide/discovering-symbols                                                                                               
  :168   Function visible_whitespace not found.                                                                                                                            
         💡 Learn more at https://phpstan.org/user-guide/discovering-symbols                                                                                               
  :180   Function visible_whitespace not found.                                                                                                                            
         💡 Learn more at https://phpstan.org/user-guide/discovering-symbols                                                                                               
  :185   Function visible_whitespace not found.                                                                                                                            
         💡 Learn more at https://phpstan.org/user-guide/discovering-symbols                                                                                               
  :196   Function visible_whitespace not found.                                                                                                                            
         💡 Learn more at https://phpstan.org/user-guide/discovering-symbols                                                                                               
  :213   Function visible_whitespace not found.                                                                                                                            
         💡 Learn more at https://phpstan.org/user-guide/discovering-symbols                                                                                               
  :219   Function visible_whitespace not found.                                                                                                                            
         💡 Learn more at https://phpstan.org/user-guide/discovering-symbols                                                                                               
  :225   Function visible_whitespace not found.                                                                                                                            
         💡 Learn more at https://phpstan.org/user-guide/discovering-symbols                                                                                               
  :231   Function visible_whitespace not found.                                                                                                                            
         💡 Learn more at https://phpstan.org/user-guide/discovering-symbols                                                                                               
  :242   Function twitter_statuses_show not found.                                                                                                                         
         💡 Learn more at https://phpstan.org/user-guide/discovering-symbols                                                                                               
  :249   Function visible_whitespace not found.                                                                                                                            
         💡 Learn more at https://phpstan.org/user-guide/discovering-symbols                                                                                               
  :266   Function twitter_media_entities not found.                                                                                                                        
         💡 Learn more at https://phpstan.org/user-guide/discovering-symbols                                                                                               
  :270   Function visible_whitespace not found.                                                                                                                            
         💡 Learn more at https://phpstan.org/user-guide/discovering-symbols                                                                                               
  :273   Function twitter_expand_entities not found.                                                                                                                       
         💡 Learn more at https://phpstan.org/user-guide/discovering-symbols                                                                                               
  :277   Function visible_whitespace not found.                                                                                                                            
         💡 Learn more at https://phpstan.org/user-guide/discovering-symbols                                                                                               
  :282   Function visible_whitespace not found.                                                                                                                            
         💡 Learn more at https://phpstan.org/user-guide/discovering-symbols                                                                                               
 ------ ------------------------------------------------------------------------------------------------------------------------------------------------------------------ 

 ------ --------------------------------------------------------------------------------------------------------- 
  Line   Module/Debug/ItemBody.php                                                                                
 ------ --------------------------------------------------------------------------------------------------------- 
  :35    Method Friendica\Module\Debug\ItemBody::content() should return string but return statement is missing.  
 ------ --------------------------------------------------------------------------------------------------------- 

 ------ --------------------------------------------------------------------------------------------------------- 
  Line   Object/Api/Mastodon/Preferences.php                                                                      
 ------ --------------------------------------------------------------------------------------------------------- 
  :42    Access to an undefined property Friendica\Object\Api\Mastodon\Preferences::$posting:default:visibility.  
         💡 Learn more: https://phpstan.org/blog/solving-phpstan-access-to-undefined-property                     
  :43    Access to an undefined property Friendica\Object\Api\Mastodon\Preferences::$posting:default:sensitive.   
         💡 Learn more: https://phpstan.org/blog/solving-phpstan-access-to-undefined-property                     
  :44    Access to an undefined property Friendica\Object\Api\Mastodon\Preferences::$posting:default:language.    
         💡 Learn more: https://phpstan.org/blog/solving-phpstan-access-to-undefined-property                     
  :45    Access to an undefined property Friendica\Object\Api\Mastodon\Preferences::$reading:expand:media.        
         💡 Learn more: https://phpstan.org/blog/solving-phpstan-access-to-undefined-property                     
  :46    Access to an undefined property Friendica\Object\Api\Mastodon\Preferences::$reading:expand:spoilers.     
         💡 Learn more: https://phpstan.org/blog/solving-phpstan-access-to-undefined-property                     
 ------ --------------------------------------------------------------------------------------------------------- 

 ------- ------------------------------------------------ 
  Line    Protocol/ActivityPub/Receiver.php               
 ------- ------------------------------------------------ 
  :1900   Cannot unset offset -1 on array<mixed, mixed>.  
 ------- ------------------------------------------------ 

 ------ --------------------------------------------------------------------------------------------------------- 
  Line   Util/Temporal.php                                                                                        
 ------ --------------------------------------------------------------------------------------------------------- 
  :347   Method Friendica\Util\Temporal::getRelativeDate() should return string but return statement is missing.  
 ------ --------------------------------------------------------------------------------------------------------- 

                                                                                                                        
 [ERROR] Found 67 errors                                                                                                
                                                                                                                        

Script phpstan analyze --memory-limit 1024M --configuration .phpstan.neon handling the phpstan event returned with error code 1
```
</details> 

Requires merge of #14526, #14539 and #14523.

Art4 committed:
install PHPStan for static code analysis

Art4 committed:
Add license note an scan directory for PHPStan

Art4 committed:
add missing return statement

Art4 committed:
Extract visible_whitespace() into separate method

Art4 committed:
Extract visible_whitespace() into separate method in ActivityPubConversion

Art4 committed:
Fix warnings in Receiver.php

Art4 committed:
add test for Preferences class

Art4 committed:
refactor of Friendica\Object\Api\Mastodon\Preferences class

Art4 committed:
Refactor Friendica\Module\Debug\ItemBody class

Art4 committed:
Remove call of non-existing functions

This functions were removed in commit a97b3f690c9e4476ba638236897507100f77af6f

- twitter_statuses_show()
- twitter_media_entities()
- twitter_expand_entities()

see https://github.com/friendica/friendica-addons/commit/a97b3f690c9e4476ba638236897507100f77af6f

Art4 committed:
Fix wrong use of static

MrPetovan commented:
It looks like this PR also includes https://github.com/friendica/friendica/pull/14523, is this intended?

Art4 commented:
> It looks like this PR also includes #14523, is this intended?

Yes, this is intended because this PR installs PHPStan using composer. If I do this with composer v1 I will have to resolve merge conflict in the `composer.lock` later.

Another reason is that I don't have to wait for the previous PRs to merge but I can already work with their improvements. This PR will stay as a draft as long the required PRs are merged. If they are rejected I will have to revert their changes from this PR.

Art4 committed:
remove unused code

this code was never called because $body is never declared
and hence the if could never be true

Art4 committed:
Remove unused code

Art4 committed:
Fix variable name

The variable was renamed but this line was accidently overseen

see https://github.com/friendica/friendica/commit/62eb16e9ad9adbd2fa712cccbb40ab53b8b16e5b#diff-785d24b46fdf749363656f33208f941ea0eca809c7c76452180a0bfc16df68a3R2555

Art4 committed:
Fix defaults in Contact model

Art4 committed:
Remove support for apc

apc is deprecated since PHP 5.5 and not compatible with PHP 7

Art4 committed:
Remove support for apc

Art4 committed:
Remove missleading template phpdoc

Art4 committed:
add missing property in FixAPDeliveryWorkerTaskParameters

Art4 committed:
Fix usage of static

Art4 committed:
Add missing return

This needs some refactoring: returning nothing implicit returns null, thas will be convert to int 0.
I return false instead of nothing to keep this behaviour for BC.

Art4 committed:
Fix returns in Console commands

Art4 committed:
add missing use statement

Art4 committed:
Fix name of Mode class

Art4 committed:
Ignore false positive PHPStan error

see https://github.com/friendica/friendica-addons/pull/1363

Art4 commented:
This PR is ready for review. Because this PR requires #14526 and #14523 one can compare the changes directly between the branches here: https://github.com/Art4/friendica/compare/improve-tests...Art4:friendica:add-static-code-analysis?expand=1

I'm not sure why the CI pipelines fails because looking at the details for the specific pipelines I only see a grey page.

Art4 commented:
While we are waiting for #14523 one can look at the relevant changes comparing the branches:

https://github.com/Art4/friendica/compare/improve-tests...Art4:friendica:add-static-code-analysis

Art4 committed:
Fix namespace

Art4 committed:
Fix cs

Art4 committed:
Move test into Unit test namespace

MrPetovan commented:
> I'm not sure why the CI pipelines fails because looking at the details for the specific pipelines I only see a grey page.

Same here, I can see a couple Javascript errors in https://ci.friendi.ca/repos/2/pipeline/2448/1 but the JS file is minified so it is impossible to figure out the actual cause.

annando commented:
Yeah, the pipeline currently is a huge problem. Possibly as some interim solution we should activate an additional pipeline on Github (like @Art4 already suggested) that does the same job. I just don't know how to do it.

Art4 commented:
I'm already trying to manually start github actions, but there is an error with mariadb.

https://github.com/Art4/friendica/actions/runs/11724292576/job/32657882187

Branch: https://github.com/Art4/friendica/tree/restore-github-actions

I will create a PR where we can investigate this issue.

Art4 committed:
Merge branch 'improve-tests' into add-static-code-analysis

Art4 committed:
Fix code style

annando commented:
@Art4 the pipeline now works again.

Art4 committed:
Merge branch 'improve-tests' into add-static-code-analysis

Art4 committed:
fix code style

Art4 committed:
add composer script to recreate lang file

Art4 committed:
recreate messages.po

Art4 commented:
I fixed the ci pipelines. After #14526 is merged this PR can be reviewed.

MrPetovan reviewed (commented):
Don't forget to rebase your branch now that the test PR was merged!

Art4 committed:
Merge branch 'develop' into add-static-code-analysis

MrPetovan commented on src/BaseCollection.php:
```diff
@@ -89,7 +89,7 @@ public function column(string $column, $index_key = null): array
 	public function map(callable $callback): BaseCollection
 	{
-		return new static(array_map($callback, $this->getArrayCopy()), $this->getTotalCount());
+		return new self(array_map($callback, $this->getArrayCopy()), $this->getTotalCount());
```
This is wrong, as it will return a `BaseCollection` instance when an inherited class instance is expected instead. If the return type hint allows for `static`, then it should be used instead.

Art4 replied:
You are right. I wasn't aware that this class would also be extended. The problem with using `static` is that the extending class can overwrite the `__construct()` method with different parameters. See https://phpstan.org/blog/solving-phpstan-error-unsafe-usage-of-new-static

I will mute this error for now by using `get_class($this);` but this has to be refactored in the future.

MrPetovan replied:
I saw this, and in one instance we are extending the constructor, so we can't use any of the remediation solutions. It's another one of these "it might break if" issues that currently don't apply to us that we unfortunately have to mute rather than "fix".

MrPetovan replied:
I'm also curious about the different between `static` and `get_class($this)` in this case, why does it mute this warning?

Art4 replied:
> I'm also curious about the different between `static` and `get_class($this)` in this case, why does it mute this warning?

I cannot say for sure but at level 3 PHPStan will start again complaining about this code.

MrPetovan commented on src/BaseCollection.php:
```diff
@@ -102,7 +102,7 @@ public function map(callable $callback): BaseCollection
 	public function filter(callable $callback = null, int $flag = 0): BaseCollection
 	{
-		return new static(array_filter($this->getArrayCopy(), $callback, $flag));
+		return new self(array_filter($this->getArrayCopy(), $callback, $flag));
```
This is wrong, as it will return a `BaseCollection` instance when an inherited class instance is expected instead. If the return type hint allows for `static`, then it should be used instead.

Art4 replied:
fixed

MrPetovan commented on src/BaseCollection.php:
```diff
@@ -112,14 +112,14 @@ public function filter(callable $callback = null, int $flag = 0): BaseCollection
 	public function reverse(): BaseCollection
 	{
-		return new static(array_reverse($this->getArrayCopy()), $this->getTotalCount());
+		return new self(array_reverse($this->getArrayCopy()), $this->getTotalCount());
```
This is wrong, as it will return a `BaseCollection` instance when an inherited class instance is expected instead. If the return type hint allows for `static`, then it should be used instead.

Art4 replied:
fixed.

MrPetovan commented on src/BaseCollection.php:
```diff
@@ -128,7 +128,7 @@ public function chunk(int $length): array
 
 		return array_map(function ($array) {
-			return new static($array);
+			return new self($array);
```
This is wrong, as it will return a `BaseCollection` instance when an inherited class instance is expected instead. If the return type hint allows for `static`, then it should be used instead.

Art4 replied:
fixed

MrPetovan commented on src/Model/Item.php:
```diff
@@ -3198,10 +3198,10 @@ public static function getPermissionsConditionArrayByUserId(int $owner_id): arra
 			// Authenticated visitor - fetch the matching permissionsets
 			$permissionSets = DI::permissionSet()->selectByContactId($remote_user, $owner_id);
-			if (!empty($set)) {
+			if (!empty($permissionSets)) {
```
Good catch!

MrPetovan commented on src/Module/Debug/ActivityPubConversion.php:
```diff
@@ -134,4 +126,9 @@ function visible_whitespace($s)
 		return $o;
 	}
+
+	private function visible_whitespace(string $s): string
```
This used to be a quick helper function used only in Debug classes. If you must make it into a full-fledged class method, please move it to the `Content\Text\HTML` class and rename it `visibleWhitespace`.

Art4 replied:
The problem here was the definition of a named function inside a method. This way you can't call the method twice, see https://3v4l.org/Y7iu7

If this quick helper function is only for debugging I would keep it as a private method. There is no problem with using a "full-fledged" method for this. I would not recommend to create a new dependency to the `Content\Text\HTML`, because this will extend our API surface that we will have to maintain. If you like I can also create an anonymous function inside the method to fix this.

MrPetovan replied:
Interesting, I didn't consider that. However the `rawContent()`, `content()` and `post()` methods are meant to be only called once per request so the issue never arose. Was it stunting the tests by any chance?

As for the dependency to `Content\Text\HTML`, it's the one area where I know how to write extensive unit tests since methods are pure and don't depend on any other service. This makes me confident about using it as a dependency itself.

If it still had been used only in one place, I would have preferred going the private function route, but since it's now used in two different places, I'd rather it being centralized in a single place which I think should be `Content\Text\HTML`.

Art4 replied:
> Was it stunting the tests by any chance?

I'm not sure that this method is called in tests but if it was called twice you definitely would see it.

> As for the dependency to `Content\Text\HTML`, it's the one area where I know how to write extensive unit tests since methods are pure and don't depend on any other service. This makes me confident about using it as a dependency itself.
>
> If it still had been used only in one place, I would have preferred going the private function route, but since it's now used in two different places, I'd rather it being centralized in a single place which I think should be `Content\Text\HTML`.

It is not about the `Content\Text\HTML` class but about the `ActivityPubConversion` class that worries me more. In my opinion this small html code would be best placed in the `view/templates/debug/activitypubconversion.tpl` template file. This should be done in a separate PR.

I will change this code to an anonymous function, do you agree?

MrPetovan replied:
Absolutely, it isn't big enough to warrant a proper function/method.

MrPetovan commented on src/Module/Debug/ItemBody.php:
```diff
@@ -11,36 +11,43 @@
 use Friendica\Core\System;
 use Friendica\DI;
 use Friendica\Model\Post;
-use Friendica\Network\HTTPException;
```
I've used this convention extensively through the code base where I only import the HTTP exceptions once, otherwise I feel like it clutters the `use` block for little benefit.

Art4 replied:
Cluttering the `use` block is a good thing because it shows you the code smell. If the `use` block is to big its an indicator that the class maybe does too much.

MrPetovan replied:
In general I'd agree, however I think exceptions are the exception. For best results, exceptions should be as unique as possible (although this can also be achieved with error codes). We already are simplifying to only HTTP-like exceptions, but there still are dozens of them, and I don't think using several of them where appropriate is a code smell.

Art4 replied:
Fully agreed. In this class imo there aren't much dependencies. But with every new used class one will get a better feeling if the class starts to getting a monolith.

Another great thing about writing every FQCN in the `use` block is that you can easily find all places where a class is used i.e. searching for `use Friendica\Network\HTTPException\NotFoundException`.

MrPetovan replied:
I see what you mean, although it's only useful for full-text search, as static analyzers within IDEs will correctly identify usage even with the relative `use` calls.

MrPetovan commented on src/Protocol/ActivityPub/Receiver.php:
```diff
@@ -1896,8 +1896,14 @@ private static function processObject(array $object, string $actor)
-		unset($object_data['reception_type'][-1]);
+
+		if (array_key_exists(-1, $object_data['receiver'])) {
+			unset($object_data['receiver'][-1]);
```
What's the rationale behind this? I thought `unset` was permissive when it comes to undefined symbols including no-existent array keys?

Art4 replied:
It makes us more aware of the array values. Yes, `unset()` is permissive for undefined symbols but we should move to more specific types. Strict types in arrays will even let PHPStan recognize in higher levels when the call of `unset()` is unnecessary.

I improved the code a little bit so we only have one check and no unnecessary calls of `unset()`.

Art4 committed:
use get_class() instead of static() in BaseCollection

Art4 committed:
Improve check for unlisted receiver

MrPetovan commented on src/Protocol/ActivityPub/Receiver.php:
```diff
@@ -1890,18 +1890,15 @@ private static function processObject(array $object, string $actor)
 		$object_data['receiver_urls']  = self::getReceiverURL($object);
 		$object_data['receiver']       = $receivers;
 		$object_data['reception_type'] = $reception_types;
+		$object_data['unlisted'] = false;
```
Please align the assignment operators.
```suggestion
		$object_data['unlisted']       = false;
```

Art4 committed:
Upgrade to PHPStan 2.0

Art4 committed:
Update src/Protocol/ActivityPub/Receiver.php

Co-authored-by: Hypolite Petovan <hypolite@mrpetovan.com>

Art4 committed:
replace method with anonymous function

MrPetovan commented on src/Protocol/ActivityPub/Receiver.php:
```diff
@@ -1890,14 +1890,20 @@ private static function processObject(array $object, string $actor)
+			$object_data['unlisted'] = true;
+			unset($object_data['receiver'][-1]);
+
+			if (array_key_exists(-1, $object_data['reception_type'])) {
```
I thought you simplified to a single condition on line 1899, why the change of heart?

Art4 replied:
Sorry, totally forgot about this: I updated to [PHPStan 2.0](https://phpstan.org/blog/phpstan-2-0-released-level-10-elephpants) and now PHPStan started complaining about this line too. :see_no_evil: 

I'll try to refactor the method again.

Art4 replied:
I've refactored the whole method to check for `-1` as key upfront. This way a value with this key will never been set and don't has to be `unset()` again. 

Having some tests for this method would be nice but It would be to much work atm to mock the other private static methods.

Art4 committed:
refactor Receiver::processObject() method

MrPetovan commented:
Waiting for release to merge if I understood correctly.

Art4 commented:
> Waiting for release to merge if I understood correctly.

@MrPetovan A release should not slow down the development. Please see #14555.
